### PR TITLE
Switch local or global val/test set

### PR DESCRIPTION
**Pull request by arthurPignet**
add an option on how compute val score and test score. If set to local, the global test/val set will be split between partners exactly like the train set. it will also be corrupted as the train set is. 

Then the val score and test score for global model will be the average of the scores computed on each partners set. 

 closing #287

### CODE REVIEW

File: mplc/scenario.py [hunk 1]
```diff
@@ -180,6 +183,15 @@ def __init__(
                 f"Computation use the full dataset for scenario #{scenario_id}"
             )
 
+        if test_set in ['local', 'global', 'both']:
+            self.test_set = test_set
+        else:
+            raise ValueError(f'Test set can be \'local\', \'global\' or \'both\', not {test_set}')
+        if test_set in ['local', 'global', 'both']:
```
Comment (bowni) [inline mplc/scenario.py:190]: Here it should be `val_set` which should be tested against the list

File: mplc/scenario.py [hunk 2]
```diff
@@ -550,19 +562,196 @@ def split_data_advanced(self, is_logging_enabled=True):
...
+
+            y_val = LabelEncoder().fit_transform([str(y) for y in self.dataset.y_val])
+
+            # Stratify the dataset into clusters per labels
+            x_val_for_cluster, y_val_for_cluster, nb_samples_per_cluster_val = {}, {}, {}
+            for label in labels:
+                idx_in_full_valset = np.where(y_val == label)
+                x_val_for_cluster[label] = self.dataset.x_val[idx_in_full_valset]
+                y_val_for_cluster[label] = self.dataset.y_val[idx_in_full_valset]
+                nb_samples_per_cluster_val[label] = len(y_val_for_cluster[label])
```
Comment (bowni) [inline mplc/scenario.py:583]: This leads to a lot lot of code duplication. Couldn't we factorize `split_data_advanced` in a function which we would apply to the datasets? Or sequence it the other way around, first the split among partners, then the val and test split?
Not sure what would be best, but we should try to avoid duplicating too much code.

**Review by bowni — CHANGES_REQUESTED** [review]

File: tests/unit_tests.py [hunk 3]
```diff
@@ -99,14 +100,15 @@ def create_Partner(create_all_datasets):
...
-                                Cifar10, ["advanced", [[4, "specific"], [6, "shared"], [4, "shared"]]],
-                                ['not-corrupted'] * 3)),
+                params=((Mnist, "random", ['not-corrupted'] * 3),
+                        (Mnist, "random", ['permutation', Redundancy(0.2), Duplication(duplicated_partner_id=0)]),
+                        (Mnist,
+                         AdvancedSplitter([0.3, 0.5, 0.2], [[4, "specific"], [6, "shared"], [4, "shared"]]),
+                         ['not-corrupted'] * 3),
+                        (Cifar10, "random", ['not-corrupted'] * 3),
+                        (Cifar10,
+                         AdvancedSplitter([0.3, 0.5, 0.2], [[4, "specific"], [6, "shared"], [4, "shared"]]),
```
Comment (natct10) [inline tests/unit_tests.py:110]: So only `AdvancedSplitter` is tested here?

Comment (arthurPignet) [reply tests/unit_tests.py:110]: No, random is tested too

Comment (arthurPignet) [reply tests/unit_tests.py:110]: But indeed we should test stratified too

File: mplc/corruption.py [hunk 4]
```diff
@@ -82,15 +110,27 @@ def generate_matrix(self):
...
+        idx_val = self.corrupted_val_index
+        for i in idx_val:
+            temp = np.zeros((self.partner.num_labels,))
+            temp[np.random.choice(self.partner.num_labels, p=self.matrix[np.argmax(self.partner.y_val[i])])] = 1
+            self.partner.y_val[i] = temp
+        logger.debug(f"   Partner #{self.partner.id}: Done.")
+        idx_test = self.corrupted_test_index
+        for i in idx_test:
+            temp = np.zeros((self.partner.num_labels,))
+            temp[np.random.choice(self.partner.num_labels, p=self.matrix[np.argmax(self.partner.y_test[i])])] = 1
```
Comment (natct10) [inline mplc/corruption.py:131]: Need coffee cc @RomainGoussault

File: mplc/doc/documentation.md [hunk 5]
```diff
@@ -232,19 +232,21 @@ There are 2 ways to select a dataset. You can either choose a pre-implemented da
...
+    - `'AdvancedSplitter'`: in certain cases it might be interesting to split the dataset among partners in a more elaborate way. For that we consider the data samples from the initial dataset as split in clusters per data labels. The advanced split is configured by indicating, for each partner in sequence, the following 2 elements: `[[nb of clusters (int), 'shared' or 'specific']]`. Practically, you can either instantiate your `AdvancedSplitter` object, and pass this list `[[nb of clusters (int), 'shared' or 'specific']]` to the keyword argument `description`, or use the string identifier and pass the list `[[nb of clusters (int), 'shared' or 'specific']]` to the scenario via the keyword argument `samples_split_configuration`.
+        String identifier:`'advanced'`.
+        Configuration:  
+        - `nb of clusters (int)`: the given partner will receive data samples from that many different clusters (clusters of data samples per labels/classes)
+        - `'shared'` or `'specific'`:
+         - `'shared'`: all partners with option `'shared'` receive data samples picked from clusters they all share data samples from
```
Comment (natct10) [inline mplc/doc/documentation.md:246]: one extra space needed

File: tests/unit_tests.py [hunk 6]
```diff
@@ -324,6 +270,157 @@ def test_generate_new_model(self, create_all_datasets):
...
+    def test_random_splitter_global(self, create_all_datasets):
+        splitter = RandomSplitter([0.1, 0.2, 0.3, 0.4])
+        dataset = create_all_datasets
+        partners_list = [Partner(i) for i in range(len(splitter.amounts_per_partner))]
+        splitter.split(partners_list, dataset)
+        for p in partners_list:
+            assert len(p.y_val) == 0, "validation set is not empty in spite of the val_set == 'global'"
+            assert len(p.y_test) == 0, "test set is not empty in spite of the val_set == 'global'"
+            assert len(p.x_train) == len(p.y_train), 'labels and samples numbers mismatches'
+            assert (p.final_nb_samples / len(dataset.y_train) - splitter.amounts_per_partner[p.id]) \
```
Comment (natct10) [inline tests/unit_tests.py:289]: Top!

File: mplc/scenario.py (hunk 2 above)
Comment (arthurPignet) [reply mplc/scenario.py]: Done with splitter object !

**Review by bowni — APPROVED** [review]

**Review by natct10 — APPROVED** [review]
Nice!
